docs/config.md
# Configuration

Application settings are in YAML file `config.yaml`

# Application configuration

`config.yaml` example:
```yaml
login: true
http_port: 0
https_port: 8000
keystore_path: 'certificate_with_certificate_chain.p12'
keystore_password: 'myPassword'
```

# Configuration YAML properties

**login**

No login needed, no access restrictions:
```yaml
login: false
```

Login needed:
```yaml
login: true
```
---

**http_port**

Application server listening on plain HTTP port.

Default port 8080:
```yaml
http_port: 8080
```

Disable HTTP, not listening on HTTP port:
```yaml
http_port: 0
```
---

**https_port**

Application server listening on encrypted secure HTTPS port. Needs keystore file at `keystore_path` and password at `keystore_password`.

Default port 8000:
```yaml
https_port: 8000
```

Disable HTTPS, not listening on HTTPS port:
```yaml
https_port: 0
```
---

**keystore_path**

*Needed for HTTPS connections only. Not needed for HTTP connections, so no keystore file. In the distribution package, no keystore file is included.*

Valid certificate for HTTPS encryption. Needs correct password at `keystore_password`.

Default `keystore.jks` filename at application root folder:
```yaml
keystore_path: 'keystore.jks'
```

Keystore should be in standardized [**PKCS #12**](https://en.wikipedia.org/wiki/PKCS_12) format, containing signed certificate, public key, encrypted private key and full certificate chain:
```yaml
keystore_path: 'certificate_with_certificate_chain.p12'
```
*See [HTTPS certificates](https_certificates.md)*.

---

**keystore_password**

*Needed for HTTPS connections only. This password opens the keystore file. This password has no relation to user login accounts.*

Password for keystore. Needed by `keystore_path`.

Setting correct password needed to decrypt the keystore file at keystore_path:
```yaml
keystore_password: 'myPassword'
```
---

docs/config_audiodb.md
# AudioDB Configuration

Application settings are in YAML file `config.yaml`

`config.yaml` example:
```yaml
login: true
http_port: 8080
    
audio:
  project: my_audio_project
  time_zone: UTC+1
  root_data_path: ./audio
  root_path: ./audio_meta
  device_inventory_file: audio_recorder_inventory.csv
  label_definitions_file: audio_label_definitions.yaml    
  
  player_fft_window: 1024
  player_fft_window_step_factor: 1
  player_time_expansion_factor: 1
  player_spectrum_threshold: 16.4
  player_fft_intensity_max: 26
  player_static_lines_frequency: [2000, 3000, 8000, 16000, 45000, 50000, 150000]
  #player_fft_cutoff_lower_frequency: 2000
  #player_fft_cutoff_upper_frequency: 10000 
  detail_fft_window_overlap_percent: 92
  
  profiles:
    birds:
      player_fft_window: 4096
      player_fft_window_step_factor: 0.125
      player_time_expansion_factor: 1
      player_spectrum_threshold: 16.4
      player_fft_intensity_max: 26
      player_static_lines_frequency: [3000, 8000]
      player_fft_cutoff_lower_frequency: 2000
      player_fft_cutoff_upper_frequency: 10000 
      detail_fft_window_overlap_percent: 92
    birds_scan:
      player_fft_window: 4096
      player_fft_window_step_factor: 1
      player_spectrum_shrink_Factor: 4
      player_time_expansion_factor: 1
      player_spectrum_threshold: 16.4
      player_fft_intensity_max: 26
      player_static_lines_frequency: [3000, 8000]
      player_fft_cutoff_lower_frequency: 2000
      player_fft_cutoff_upper_frequency: 10000 
      detail_fft_window_overlap_percent: 92
```

src/audio/Config.java
package audio;


import org.tinylog.Logger;

import photo.PhotoConfig;
import util.collections.ReadonlyList;
import util.collections.vec.Vec;
import util.yaml.YamlMap;

public class Config {
	
	public static final Config DEFAULT = new Config();

	public final boolean login;
	public final Account default_account;	
	public final ReadonlyList<JwsConfig> jwsConfigs;
	public final int http_port;
	public final int https_port;
	public final String keystore_path;
	public final String keystore_password;
	//public final AudioConfig audioConfig;
	public final AudioProjectConfig audioConfig;
	public final PhotoConfig photoConfig;
	
	@SuppressWarnings("unchecked")
	private Config() {
		login = true;
		default_account = Account.ofPassword("anonymous", "", "", new String[] {"admin"});
		this.jwsConfigs = ReadonlyList.EMPTY;
		this.http_port = 8080;
		this.https_port = 8000;
		this.keystore_path = "keystore.jks";
		this.keystore_password = "";
		//this.audioConfig = AudioConfig.ofYAML(YamlMap.EMPTY_MAP);
		this.audioConfig = AudioProjectConfig.DEFAULT;
		this.photoConfig = PhotoConfig.ofYAML(YamlMap.EMPTY_MAP);
	}
	
	public Config(boolean login, Account default_account, ReadonlyList<JwsConfig> jwsConfigs, int http_port, int https_port, String keystore_path, String keystore_password, AudioProjectConfig audioConfig, PhotoConfig photoConfig) {
		this.login = login;
		this.default_account = default_account;
		this.jwsConfigs = jwsConfigs;
		this.http_port = http_port;
		this.https_port = https_port;
		this.keystore_path = keystore_path;
		this.keystore_password = keystore_password;
		this.audioConfig = audioConfig;
		this.photoConfig = photoConfig;
	}
	
	public static Config ofYAML(YamlMap yamlMap) {
		boolean login = yamlMap.optBoolean("login", DEFAULT.login);
		Account default_account = DEFAULT.default_account;
		int http_port = yamlMap.optInt("http_port", DEFAULT.http_port);
		int https_port = yamlMap.optInt("https_port", DEFAULT.https_port);
		String keystore_path = yamlMap.optString("keystore_path", DEFAULT.keystore_path);
		String keystore_password = yamlMap.optString("keystore_password", DEFAULT.keystore_password);
		
		Vec<JwsConfig> jwsConfigs = new Vec<JwsConfig>();
		
		try {
			for(YamlMap yamlJws:yamlMap.optList("jws").asMaps()) {
				try {
					JwsConfig jwsConfig = JwsConfig.ofYAML(yamlJws);
					jwsConfigs.add(jwsConfig);
				} catch (Exception e) {
					Logger.warn(e);
				}
			}
		} catch (Exception e) {
			Logger.warn(e);
		}
		
		//AudioConfig audioConfig = AudioConfig.ofYAML(yamlMap.optMap("audio"));
		AudioProjectConfig audioConfig = new AudioProjectConfig(new AudioProjectConfig.Builder(yamlMap.optMap("audio")));
		PhotoConfig photoConfig = PhotoConfig.ofYAML(yamlMap.optMap("photo"));
		
		return new Config(login, default_account, jwsConfigs.readonlyWeakView(), http_port, https_port, keystore_path, keystore_password, audioConfig, photoConfig);
	}
	
	public boolean hasHTTPport() {
		return http_port > 0;
	}
	
	public boolean hasHTTPSport() {
		return https_port > 0;
	}
	
	public boolean enableHttps() {
		return hasHTTPSport() && !keystore_password.isEmpty();
	}
}
